File: 2023/11/11.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strings"
)

type G struct{ row, col int }

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func solve(gs []G, k int) int {
	gs = slices.Clone(gs)
	nr, nc := 0, 0
	grows, gcols := map[int]bool{}, map[int]bool{}
	for _, g := range gs {
		if g.col+1 > nc {
			nc = max(g.col+1, nc)
		}
		if g.row+1 > nr {
			nr = max(g.row+1, nr)
		}
		grows[g.row] = true
		gcols[g.col] = true
	}
	emptyrows, emptycols := []int{}, []int{}
	for r := 0; r < nr; r++ {
		if !grows[r] {
			emptyrows = append(emptyrows, r)
		}
	}
	for c := 0; c < nc; c++ {
		if !gcols[c] {
			emptycols = append(emptycols, c)
		}
	}
	slices.Sort(emptyrows)
	slices.Sort(emptycols)
	slices.Reverse(emptyrows)
	slices.Reverse(emptycols)

	for _, er := range emptyrows {
		for i := range gs {
			if gs[i].row > er {
				gs[i].row += k - 1
			}
		}
	}
	for _, ec := range emptycols {
		for i := range gs {
			if gs[i].col > ec {
				gs[i].col += k - 1
			}
		}
	}

	s := 0
	for i, g1 := range gs {
		for _, g2 := range gs[i+1:] {
			s += abs(g1.row-g2.row) + abs(g1.col-g2.col)
		}
	}
	return s
}

func main() {
	file, err := os.ReadFile("11.txt")
	if err != nil {
		panic(err)
	}
	input := string(file)
	lines := strings.Split(input, "\n")
	gs := []G{}
	for r, line := range lines {
		for c, ch := range line {
			if ch == '#' && !slices.Contains(gs, G{r, c}) {
				gs = append(gs, G{r, c})
			}
		}
	}
	fmt.Println(solve(gs, 2))
	fmt.Println(solve(gs, 1000000))
}

File: 2023/11/11.py
from itertools import combinations

inp = open('11.txt').read()
k = 10**6

gs = {(r, c) for r, L in enumerate(inp.split("\n")) for c, ch in enumerate(L) if ch == '#'}

nr = max(r for r, _ in gs) + 1
nc = max(c for _, c in gs) + 1

grows = {r for r, _ in gs}
gcols = {c for _, c in gs}

emptyrows = sorted(r for r in range(nr) if r not in grows)
emptycols = sorted(c for c in range(nc) if c not in gcols)

for er in reversed(emptyrows):
    gs = {(r+k-1 if r > er else r, c) for r, c in gs}
for ec in reversed(emptycols):
    gs = {(r, c+k-1 if c > ec else c) for r, c in gs}

print(sum(abs(r1 - r2) + abs(c1 - c2) for (r1, c1), (r2, c2) in combinations(gs, 2)))
